flow/ThreadHelper.actor.h

```diff
@@ -600,26 +623,24 @@ Future<Void> doOnMainThread(Future<Void> signal, F f, ThreadSingleAssignmentVar<
 	return Void();
 }
 
-ACTOR template <class F>
-void doOnMainThreadVoid(Future<Void> signal, F f, Error* err) {
-	wait(signal);
-	if (err && err->code() != invalid_error_code)
-		return;
-	try {
-		f();
-	} catch (Error& e) {
-		if (err)
-			*err = e;
-	}
-}
+}  // namespace internal_thread_helper
 
+// `onMainThread` runs a functor returning a `Future` on the main thread, waits for the future, and sends either the
+// value returned from the waited `Future` or an error through the `ThreadFuture` returned from the function call.
+//
+// A workaround for cases where your functor returns a non-`Future` value is to wrap the value in an immediately
+// filled `Future`. In cases where the functor returns void, a workaround is to return a `Future<bool>(true)` that
+// can be waited on.
+//
+// TODO: Add SFINAE overloads for functors returning void or a non-Future type.
 template <class F>
 ThreadFuture<decltype(std::declval<F>()().getValue())> onMainThread(F f) {
 	Promise<Void> signal;
 	auto returnValue = new ThreadSingleAssignmentVar<decltype(std::declval<F>()().getValue())>();
 	returnValue->addref(); // For the ThreadFuture we return
-	Future<Void> cancelFuture =
-	    doOnMainThread<decltype(std::declval<F>()().getValue()), F>(signal.getFuture(), f, returnValue);
+	                       // TODO: Is this cancellation logic actually needed?
```

Comment:
This got formatted weirdly, maybe it needs a line break for clang-format not to get confused.

```suggestion
	// TODO: Is this cancellation logic actually needed?
```